src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java

```diff
@@ -0,0 +1,396 @@
+            if (partitionIndexes.length == 0) {
+                // interval does not overlap any partition - skip it
+                continue;
+            }
+            Arrays.sort(partitionIndexes);
+            int startIndex = partitionIndexes[0];
+            int endIndex = partitionIndexes[partitionIndexes.length - 1];
+            indexedIntervals.add(new PartitionLocatable<I>(startIndex, interval));
+            if (endIndex > maxEndPartitionIndexes.get(startIndex)) {
+                maxEndPartitionIndexes.set(startIndex, endIndex);
+            }
+        }
+
+        JavaRDD<R> coalescedRdd = coalesce(locatables, locatableClass, new RangePartitionCoalescer(maxEndPartitionIndexes));
+
+        // Create an RDD of intervals with the same number of partitions as the locatables, and where each interval
+        // is in its start partition.
+        JavaRDD<I> intervalsRdd = ctx.parallelize(indexedIntervals)
+                .mapToPair(interval ->
+                        new Tuple2<>(interval.getPartitionIndex(), interval.getLocatable()))
+                .partitionBy(new KeyPartitioner(locatables.getNumPartitions())).values();
+
+        // zipPartitions on coalesced read partitions and intervals, and apply the function f
+        return coalescedRdd.zipPartitions(intervalsRdd, f);
+    }
+
+    /**
+     * Turn a pair of iterators over intervals and locatables, into a single iterator over pairs made up of an interval and
+     * the locatables that overlap it. Intervals with no overlapping locatables are dropped.
+     */
+    static <R extends Locatable, I extends Locatable> Iterator<Tuple2<I, Iterable<R>>> locatablesPerShard(Iterator<R> locatables, Iterator<I> shards, SAMSequenceDictionary sequenceDictionary, int maxLocatableLength) {
+        if (!shards.hasNext()) {
+            return Collections.emptyIterator();
+        }
+        PeekingIterator<R> peekingLocatables = Iterators.peekingIterator(locatables);
+        PeekingIterator<I> peekingShards = Iterators.peekingIterator(shards);
+        Iterator<Tuple2<I, Iterable<R>>> iterator = new AbstractIterator<Tuple2<I, Iterable<R>>>() {
+            // keep track of current and next, since locatables can overlap two shards
+            I currentShard = peekingShards.next();
+            I nextShard = peekingShards.hasNext() ? peekingShards.next() : null;
+            List<R> currentLocatables = Lists.newArrayList();
+            List<R> nextLocatables = Lists.newArrayList();
+
+            @Override
+            protected Tuple2<I, Iterable<R>> computeNext() {
+                if (currentShard == null) {
+                    return endOfData();
+                }
+                while (peekingLocatables.hasNext()) {
+                    if (toRightOf(currentShard, peekingLocatables.peek(), sequenceDictionary)) {
+                        break;
+                    }
+                    R read = peekingLocatables.next();
```

Comment:
Still some code in here and below that uses the term "read" -- should generalize (eg., `currentLocatable`)